Flames/flames_drv.py

```diff
@@ -293,78 +295,125 @@ def checkSequence(self, firingSequence):
+
+        sequenceName = msgObj["name"]
+        sequence = self.pattern_manager.getPattern(sequenceName)
+
+        duration = sequence["duration"]
+        events = sequence["events"]
+        firstFiringTime = time.time()
+
+        if not self.isStopped:
+            for event in events:
+                ids = events["ids"]
+                startTime = firstFiringTime + event["startTime"]
+                endTime = startTime + event["duration"]
+
+                addresses = [pooferMapping[a] for a in ids].sort()
+                bangCommandList = makeBangCommandList(addresses)
+
+                pooferEvent = {}
+                pooferEvent["sequence"] = sequenceName
+                pooferEvent["time"] = startTime
+                pooferEvent["bangCommandList"] = bangCommandList["on"]
+
+                endPooferEvent = {}
+                endPooferEventendPooferEvent["sequence"] = sequenceName
+                endPooferEvent["time"] = endTime
+                endPooferEvent["bangCommandList"] = bangCommandList["off"]
+
+                # TODO: need to figure out best way to sort this thing
+                self.pooferEvents.append(pooferEvent)
+                self.pooferEvents.append(endPooferEvent)
+
+    def makeBangCommandList(addresses):
+        # creates a dictionary with the key being a controller ID (two digits),
+        # and values being all the channels for a given controller.
+        # returns an object with bang commands to turn poofers both on and off
+
+        onBangCommands = []
+        offBangCommands = []
 
         try:
-    
-            #steps: 
-            #check to see if electronic killswitch is enabled, if so then do nothing
-            #iterating through firingSequence, for each step:
-            #	generate eventsList
-            #	for each in eventsList:
-            #		each item becomes a new item in bangCommandList:
-            #		bangCommandList[index]="b"+(string: translate alphas to address via pooferMappings)+"t"+str(digits)
-            #		make sure that there is a 1:1 match between each step in firingSequence and each step in bangCommandList
-        
-            #translate firingSequence into bang code
-            for step in firingSequence:
-                #divide parallel firings into single events
-                parallelEvents=step.split("&")
-                #write the events into a sequence of bang commands
-                
-
-                            
-                            
-            
-                    #bangCommandList=
-                #
-            
-            
-            
-            #else:
-                #end the poof string with the "end" command.
-                #bangCommandList= + "."
-            
+            controllerDict = defaultdict(list)
+            for controllerId in address: controllerDict[controllerId[:2]].append(controllerId[2])
+
+            for i in controllerDict.keys():
+                onBangCommands.append(
```

Comment:
Currently, I'm creating one command to turn the poofer on, and one to turn it off when the duration is over. I'll just change it to let the poofers run their course then?